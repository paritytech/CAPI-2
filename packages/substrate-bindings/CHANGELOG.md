# Changelog

## Unreleased

### Breaking

- Move the `is` and `as` methods from enum values to `Enum`: `Enum.is(value, tag)`, `Enum.as(value, tag)`.

## 0.3.0 - 2024-04-25

### Added

- New type `EnumVariant<T, K>` to select one specific variant from an enum.
- Improved Enum type inference so that they can be assigned between types as long as they are compatible.

### Breaking

- The generic for `Enum<T>` is now an object of `{ [type: string]: any }`, rather than a union of `{ type: string, value: any } | ... | { type: string, value: any }`

## 0.2.0 - 2024-04-23

### Added

- Support for metadata V14

## 0.1.0 - 2024-04-22

### Changed

- Added support for `AccountId20`

## 0.0.1 - 2024-04-03

### Changed

Initial release
