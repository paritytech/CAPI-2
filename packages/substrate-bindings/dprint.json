{
  "incremental": true,
  "indentWidth": 2,
  "lineWidth": 80,
  "typescript": {
    "trailingCommas": "onlyMultiLine",
    "quoteProps": "asNeeded",
    "arrowFunction.useParentheses": "force",
    "semiColons": "asi"
  },
  "markdown": {
    "textWrap": "always"
  },
  "includes": ["**.{json,md,ts}"],
  "excludes": ["dist", "package-lock.json"],
  "plugins": [
    "https://plugins.dprint.dev/typescript-0.84.4.wasm",
    "https://plugins.dprint.dev/json-0.17.2.wasm",
    "https://plugins.dprint.dev/markdown-0.15.2.wasm"
  ]
}
