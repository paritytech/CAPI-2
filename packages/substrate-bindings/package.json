{
  "name": "@polkadot-api/substrate-bindings",
  "version": "0.6.0",
  "author": "Josep M Sobrepere (https://github.com/josepot)",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/polkadot-api/polkadot-api.git"
  },
  "license": "MIT",
  "sideEffects": false,
  "exports": {
    ".": {
      "node": {
        "production": {
          "import": "./dist/esm/index.mjs",
          "require": "./dist/min/index.js",
          "default": "./dist/index.js"
        },
        "import": "./dist/esm/index.mjs",
        "require": "./dist/index.js",
        "default": "./dist/index.js"
      },
      "module": "./dist/esm/index.mjs",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/index.js",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.js",
  "module": "./dist/esm/index.mjs",
  "browser": "./dist/esm/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --noEmit && rollup -c ../../rollup.config.js",
    "test": "vitest",
    "lint": "prettier --check README.md \"src/**/*.{js,jsx,ts,tsx,json,md}\"",
    "format": "prettier --write README.md \"src/**/*.{js,jsx,ts,tsx,json,md}\"",
    "prepack": "pnpm run build"
  },
  "dependencies": {
    "@noble/hashes": "^1.3.1",
    "@scure/base": "^1.1.1",
    "@polkadot-api/utils": "workspace:*",
    "scale-ts": "^1.6.0"
  },
  "devDependencies": {
    "bigint-conversion": "^2.4.1",
    "hash-wasm": "^4.9.0"
  }
}
