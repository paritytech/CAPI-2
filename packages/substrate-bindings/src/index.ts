export * from "./codecs"
export * from "./hashes"
export * from "./storage"
export * from "./descriptors"
