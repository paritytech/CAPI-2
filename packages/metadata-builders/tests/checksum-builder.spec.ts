import ksm from "./ksm.json"
import { V14Lookup, V15 } from "@polkadot-api/substrate-bindings"
import { expect, describe, it } from "vitest"
import { getChecksumBuilder } from "@/."

describe("getChecksumBuilder snapshots", () => {
  const builder = getChecksumBuilder(ksm as any)
  it("batched call", () => {
    const result = builder.buildCall("Utility", "batch")
    expect(result).toMatchSnapshot()
  })

  it("felloship referenda submit", () => {
    const result = builder.buildCall("FellowshipReferenda", "submit")
    expect(result).toMatchSnapshot()
  })
})

describe("buildDefinition properties", () => {
  const builder = getChecksumBuilder({ lookup } as any)

  const expectEqual = (
    a: V14Lookup[number]["def"],
    b: V14Lookup[number]["def"],
  ) => {
    const aId = lookupPush(a)
    const bId = lookupPush(b)
    expect(builder.buildDefinition(aId)).toEqual(builder.buildDefinition(bId))
  }
  const expectNotEqual = (
    a: V14Lookup[number]["def"],
    b: V14Lookup[number]["def"],
  ) => {
    const aId = lookupPush(a)
    const bId = lookupPush(b)
    expect(builder.buildDefinition(aId)).not.toEqual(
      builder.buildDefinition(bId),
    )
  }

  it("gives the same result regardless of entry point", () => {
    const referenceBuilder = getChecksumBuilder(ksm as V15)
    ksm.lookup.map((x) => x.id).forEach(referenceBuilder.buildDefinition)

    ksm.lookup.forEach((x) => {
      const secondBuilder = getChecksumBuilder(ksm as V15)
      const referenceResult = referenceBuilder.buildDefinition(x.id)
      const secondResult = secondBuilder.buildDefinition(x.id)

      expect(referenceResult).toEqual(secondResult)
    })
  })

  it("generates the same checksum if the runtime type is equal", () => {
    expect(builder.buildDefinition(knownIds.u8)).toEqual(
      builder.buildDefinition(knownIds.u32),
    )

    const compactId = lookupPush({ tag: "compact", value: knownIds.u8 })
    expect(builder.buildDefinition(compactId)).toEqual(
      builder.buildDefinition(knownIds.u32),
    )
  })

  it("generates a different checksum if the runtime type is different, for a numeric primitive", () => {
    expect(builder.buildDefinition(knownIds.u32)).not.toEqual(
      builder.buildDefinition(knownIds.u64),
    )

    const compactId = lookupPush({ tag: "compact", value: knownIds.u64 })
    expect(builder.buildDefinition(knownIds.u32)).not.toEqual(
      builder.buildDefinition(compactId),
    )
  })

  it("distinguishes between sequence and arrays", () => {
    expectNotEqual(
      { tag: "sequence", value: 0 },
      { tag: "array", value: { len: 3, type: 0 } },
    )
  })

  it("distinguishes between arrays of different length", () => {
    expectNotEqual(
      { tag: "array", value: { len: 3, type: 0 } },
      { tag: "array", value: { len: 5, type: 0 } },
    )
  })

  it("distinguishes between arrays of different type", () => {
    expectNotEqual(
      { tag: "array", value: { len: 3, type: 0 } },
      { tag: "array", value: { len: 3, type: 2 } },
    )
  })

  it("gives the same checksum for flipped tuple parameters if they have the same type", () => {
    expectEqual(
      { tag: "tuple", value: [0, 1, 2] },
      { tag: "tuple", value: [1, 0, 2] },
    )
  })

  it("distinguishes flipped tuple parameters across multiple types", () => {
    expectNotEqual(
      { tag: "tuple", value: [0, 1, 2] },
      { tag: "tuple", value: [0, 2, 1] },
    )
  })

  it("gives the same checksum for reordered struct types", () => {
    expectEqual(
      {
        tag: "composite",
        value: [
          createCompositeEntry({
            name: "foo",
            type: knownIds.u8,
          }),
          createCompositeEntry({
            name: "bar",
            type: knownIds.u64,
          }),
        ],
      },
      {
        tag: "composite",
        value: [
          createCompositeEntry({
            name: "bar",
            type: knownIds.u64,
          }),
          createCompositeEntry({
            name: "foo",
            type: knownIds.u8,
          }),
        ],
      },
    )
  })

  it("gives a different checksum for reordered tuple types", () => {
    expectNotEqual(
      {
        tag: "composite",
        value: [
          createCompositeEntry({
            type: knownIds.u8,
          }),
          createCompositeEntry({
            type: knownIds.u64,
          }),
        ],
      },
      {
        tag: "composite",
        value: [
          createCompositeEntry({
            type: knownIds.u64,
          }),
          createCompositeEntry({
            type: knownIds.u8,
          }),
        ],
      },
    )
  })

  it("gives the same checksum when there are circular references", () => {
    const aId = lookupPush({
      tag: "composite",
      value: [
        createCompositeEntry({
          name: "id",
          type: knownIds.u32,
        }),
        createCompositeEntry({
          name: "nextA",
          type: 0,
        }),
      ],
    })
    const optionId = lookupPush({
      tag: "variant",
      value: [
        {
          docs: [],
          index: 0,
          name: "Some",
          fields: [
            createCompositeEntry({
              type: aId,
            }),
          ],
        },
        {
          docs: [],
          index: 1,
          name: "None",
          fields: [],
        },
      ],
    })
    const aDef = lookup[aId].def
    if (aDef.tag === "composite") {
      aDef.value[1].type = optionId
    }

    const secondBuilder = getChecksumBuilder({ lookup } as any)

    const aChecksum = builder.buildDefinition(aId)
    const optionChecksum = builder.buildDefinition(optionId)

    const secondOptionChecksum = secondBuilder.buildDefinition(optionId)
    const secondAChecksum = secondBuilder.buildDefinition(aId)

    expect(aChecksum).toEqual(secondAChecksum)
    expect(optionChecksum).toEqual(secondOptionChecksum)
    expect(aChecksum).not.toEqual(optionChecksum)
  })

  it("gives the same checksum when there are circular references after multiple levels", () => {
    const aId = lookupPush({
      tag: "sequence",
      value: 0,
    })
    const bId = lookupPush({
      tag: "sequence",
      value: 0,
    })
    const cId = lookupPush({
      tag: "composite",
      value: [
        createCompositeEntry({
          name: "foo",
          type: 0,
        }),
        createCompositeEntry({
          name: "a",
          type: aId,
        }),
      ],
    })
    const dId = lookupPush({
      tag: "sequence",
      value: aId,
    })
    lookup[aId].def.value = bId
    lookup[bId].def.value = cId
    function perms(xs: number[]): number[][] {
      if (!xs.length) return [[]]
      return xs.flatMap((x) => {
        return perms(xs.filter((v) => v !== x)).map((vs) => [x, ...vs])
      })
    }

    const orderings = perms([aId, bId, cId, dId])
    const runOrdering = (ordering: number[]) => {
      const builder = getChecksumBuilder({ lookup } as any)
      return ordering
        .map((id) => [id, builder.buildDefinition(id)] as const)
        .sort(([a], [b]) => a - b)
    }

    const reference = runOrdering(orderings[0])

    orderings
      .slice(1)
      .forEach((ordering, i) =>
        expect(
          runOrdering(ordering),
          `checksum mismatch on case ${orderings[i + 1]}`,
        ).toEqual(reference),
      )
  })
})

const createEntry = (
  id: number,
  def: V14Lookup[number]["def"],
): V14Lookup[number] => ({
  id,
  docs: [],
  path: [],
  params: [],
  def,
})

const knownIds = {
  u8: 0,
  u32: 1,
  u64: 2,
}

const lookup: V14Lookup = [
  createEntry(knownIds.u8, {
    tag: "primitive",
    value: { tag: "u8", value: undefined },
  }),
  createEntry(knownIds.u32, {
    tag: "primitive",
    value: { tag: "u32", value: undefined },
  }),
  createEntry(knownIds.u64, {
    tag: "primitive",
    value: { tag: "u64", value: undefined },
  }),
]

const lookupPush = (def: V14Lookup[number]["def"]) => {
  const id = lookup.length
  lookup.push(createEntry(id, def))
  return id
}

const createCompositeEntry = <
  T extends Partial<{
    name: string | undefined
    type: number
    typeName: string | undefined
    docs: string[]
  }>,
>(
  value: T,
) => ({
  docs: [],
  typeName: "",
  name: undefined,
  ...value,
})
