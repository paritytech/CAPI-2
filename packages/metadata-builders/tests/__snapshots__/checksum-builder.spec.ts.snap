// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getChecksumBuilder > batched call 1`] = `"fjitig6de921r"`;

exports[`getChecksumBuilder > felloship referenda submit 1`] = `"5v8qtc0v9ju6o"`;
